nexus/src/cidata.rs

```diff
@@ -0,0 +1,93 @@
+use crate::db::{identity::Resource, model::Instance};
+use fatfs::{FileSystem, FormatVolumeOptions, FsOptions};
+use omicron_common::api::external::Error;
+use serde::Serialize;
+use std::io::{Cursor, Write};
+use uuid::Uuid;
+
+impl Instance {
+    pub fn generate_cidata(&self) -> Result<Vec<u8>, Error> {
+        // cloud-init meta-data is YAML, but YAML is a strict superset of JSON.
+        let meta_data = serde_json::to_vec(&MetaData {
+            instance_id: self.id(),
+            local_hostname: &self.runtime().hostname,
+            public_keys: "", // TODO
+        })
+        .map_err(|_| Error::internal_error("failed to serialize meta-data"))?;
+        let cidata =
+            build_vfat(&meta_data, &self.user_data).map_err(|err| {
+                Error::internal_error(&format!(
+                    "failed to create cidata volume: {}",
+                    err
+                ))
+            })?;
+        Ok(cidata)
+    }
+}
+
+#[derive(Serialize)]
+#[serde(rename_all = "kebab-case")]
+struct MetaData<'a> {
+    instance_id: Uuid,
+    local_hostname: &'a str,
+    public_keys: &'a str,
+}
+
+fn build_vfat(meta_data: &[u8], user_data: &[u8]) -> std::io::Result<Vec<u8>> {
+    // requires #![feature(int_roundings)].
+    // https://github.com/rust-lang/rust/issues/88581
+    let file_sectors = meta_data.len().unstable_div_ceil(512)
+        + user_data.len().unstable_div_ceil(512);
+    // this was reverse engineered by making the numbers go lower until the
+    // code failed (usually because of low disk space). this only works for
+    // FAT12 filesystems
+    let sectors = 42.max(file_sectors + 35 + ((file_sectors + 1) / 341 * 2));
```

Comment:
what do the numbers in this calculation represent? my FAT* filesystem ignorant read is "the lowest number of sectors we support is 42, and otherwise we have to compute enough space for FAT* formatting / metadata plus our file data", but that implies that FAT* needs 35 sectors for something?

Reply:
I actually have no idea because the only reference I could easily find is [\[\[Design of the FAT file system\]\]](https://en.wikipedia.org/wiki/Design_of_the_FAT_file_system), which I repeatedly tried to read and my eyes kept falling off the page. I wrote some code using fatfs to find out what the smallest numbers are that work with it.

This more or less spells out to:
- No device less than 42 sectors can be formatted as FAT12
- Given the total number of sectors taken up by your files, 35 are overhead
- ... plus an additional 2 sectors of overhead based on some linear relationship that has numbers I don't understand in it (why is it 341???).

Reply:
Though reading your comment earlier, it seems like you experimentally noticed the minimum FS size was 42, no? Any idea where that's coming from?

Reply:
No idea where the 42 is coming from, no. But if you try to give fatfs fewer than 42 sectors, it gives you an "unfortunate disk size" error.